loki/transform/transform_array_indexing.py

```diff
@@ -428,3 +428,66 @@ def promote_nonmatching_variables(routine, promotion_vars_dims, promotion_vars_i
     for (index, size), var_names in index_size_var_map.items():
         promote_variables(routine, var_names, -1, index=index, size=size)
     info('%s: promoted variable(s): %s', routine.name, ', '.join(promotion_vars_dims.keys()))
+
+
+def demote_variables(routine, variable_names, dimensions):
+    """
+    Demote a list of array variables by removing any occurence of a
+    provided set of dimension symbols.
+
+    Parameters
+    ----------
+    routine : :any:`Subroutine`
+        The subroutine in which the variables should be promoted.
+    variable_names : list of str
+        The names of variables to be promoted. Matching of variables against
+        names is case-insensitive.
+    dimensions : :py:class:`pymbolic.Expression` or tuple
+        Symbol name or tuple of symbol names representing the dimension
+        to remove from all occurances of the named variables.
+    """
+    dimensions = as_tuple(dimensions)
+
+    # Compare lower-case only, since we're not comparing symbols
+    vnames = tuple(name.lower() for name in variable_names)
+
+    variables = FindVariables(unique=False).visit(routine.ir)
+    variables = tuple(v for v in variables if v.name.lower() in vnames)
+
+    # Record original array shapes
+    shape_map = CaseInsensitiveDict({v.name: v.shape for v in variables})
```

Comment:
This will probably fail if `variables` includes non-Array variables. I think the utility should handle this gracefully and discard any scalar symbols in `variables`. Would be good to verify this also in a test, e.g. by adding just a scalar to the list of names passed to the utility and checking that nothing has happened to it.